### PR TITLE
WIP: fix balanceOf

### LINKED ISSUE #3089: BalanceOf() returns 1 for previous key owner.
**Describe the bug**
I may have uncovered a bug in our Lock Contract, where after transferring a key to another address, the previous owner still has a balance of `1` as returned by `balanceOf`

**To Reproduce**
 1.) For a new lock, address1 buys a key. // tokenID = 1
 2.) address1 transfers the key (with `transferFrom`) to address2.
 3.) `ownerOf(1)` // returns address2
 4.)  `balanceOf(address1)` // returns 1
 5.) `balanceOf(address2)` // returns 1

NOTE: etherscan shows that address1 still owns the token. This makes me think that the transfer itself is flawed...

**Expected behavior**
`balanceOf(address1)` // should return 0.
etherscan should show the transfer to address2.

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by nfurfaro**
# Description

This is just a start in the right direction. Still have some work to do to get tests passing again.

# Issues

Fixes #3089 

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [ ] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [ ] My code follows the style guidelines of this project, including naming conventions
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: smart-contracts/contracts/mixins/MixinKeys.sol [hunk 1]
```diff
@@ -102,12 +102,13 @@ contract MixinKeys is
   function balanceOf(
     address _owner
   )
-    external
+    public
```
Comment (nfurfaro) [inline smart-contracts/contracts/mixins/MixinKeys.sol]: I modified this to reuse `balanceOf` but think I'll change it back.

File: smart-contracts/test/Lock/erc721/transferFrom.js [hunk 2]
```diff
@@ -225,6 +221,7 @@ contract('Lock / erc721 / transferFrom', accounts => {
           from,
         })
         ID = await locks['FIRST'].getTokenIdFor.call(from)
+        console.log(ID)
```
Comment (nfurfaro) [inline smart-contracts/test/Lock/erc721/transferFrom.js]: @hardlydifficult ID at this point is 0, but should actually be a valid tokenID!

**Comment by nfurfaro** [comment]
@hardlydifficult I'm stuck on this atm...here's the latest build log. https://circleci.com/gh/unlock-protocol/unlock/46868?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link As you can see, 2 failing tests. The second one is `Uncaught Error: Returned values aren't valid, did it run Out of Gas?
      at ABICoder.decodeParameters (node_modules/truffle/build/webpack:/~/web3-eth-abi/src/index.js:226:1)` which shows up in a thread about a web3 error : https://github.com/ethereum/web3.js/issues/1916#issuecomment-427398031

I need to take a break from this, it's killing me. If you know what's going on here off the top of your head, well that would be swell.

**Comment by hardlydifficult** [comment]
That second error, 'out of gas', is because of the first error.  So that's the one to focus on.  hmm I'm not sure which line is failing though - wish the stack trace didn't suck =p.

Maybe add some console logs to narrow it down

**Comment by nfurfaro** [comment]
Yeah the stack trace does suck. Ive been playing with adding console.logs, as well as commenting out lines, to try to isolate the failure. Will keep at it after a break.

**Comment by hardlydifficult** [comment]
@nfurfaro fixed the gas report - sorry for the trouble.  not exactly sure how it was working previously ;)

**Comment by nfurfaro** [comment]
@hardlydifficult Thanks! I wasn't sure what was causing that error, but happy to move on!

**Pull request merged**